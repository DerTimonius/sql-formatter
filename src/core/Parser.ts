import { EOF_TOKEN, type Token, TokenType } from './token';
/* eslint-disable no-cond-assign */

export type Statement = {
  type: 'statement';
  children: AstNode[];
};

// Wrapper for plain nodes inside AST
export type TokenNode = {
  type: 'token';
  token: Token;
};

export type ParenthesizedExpr = {
  type: 'parenthesized_expr';
  children: AstNode[];
};

export type AstNode = ParenthesizedExpr | TokenNode;

export const isTokenNode = (node: AstNode): node is TokenNode => node.type === 'token';

/**
 * A rudimentary parser that slices token stream into list of SQL statements.
 */
export default class Parser {
  private index = 0;

  constructor(private tokens: Token[]) {}

  public parse(): Statement[] {
    const statements: Statement[] = [];
    let stat: Statement | undefined;
    while ((stat = this.statement())) {
      statements.push(stat);
    }
    return statements;
  }

  private statement(): Statement | undefined {
    const tokens: TokenNode[] = [];
    while (true) {
      if (this.look().value === ';') {
        tokens.push({ type: 'token', token: this.next() });
        return { type: 'statement', children: tokens };
      } else if (this.look().type === TokenType.EOF) {
        if (tokens.length > 0) {
          return { type: 'statement', children: tokens };
        } else {
          return undefined;
        }
      } else {
        tokens.push({ type: 'token', token: this.next() });
      }
    }
  }

  // Returns current token without advancing the pointer
  private look(): Token {
    return this.tokens[this.index] || EOF_TOKEN;
  }

  // Returns current token and advances the pointer to next token
  private next(): Token {
    return this.tokens[this.index++] || EOF_TOKEN;
  }
}
