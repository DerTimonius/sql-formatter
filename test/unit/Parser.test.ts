import Parser from 'src/parser/Parser';
import Tokenizer from 'src/lexer/Tokenizer';

describe('Parser', () => {
  const parse = (sql: string) => {
    const tokens = new Tokenizer({
      reservedCommands: ['SELECT', 'SELECT DISTINCT', 'FROM', 'WHERE', 'LIMIT', 'CREATE TABLE'],
      reservedDependentClauses: ['WHEN', 'ELSE'],
      reservedSetOperations: ['UNION'],
      reservedJoins: ['JOIN'],
      reservedKeywords: ['BETWEEN', 'LIKE', 'ON', 'USING'],
      reservedFunctionNames: ['SQRT', 'OFFSET'],
      openParens: ['(', '['],
      closeParens: [')', ']'],
      stringTypes: ["''"],
      identTypes: ['""'],
    }).tokenize(sql);
    return new Parser(tokens).parse();
  };

  it('parses empty list of tokens', () => {
    expect(parse('')).toEqual([]);
  });

  it('parses multiple statements', () => {
    expect(parse('foo; bar')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "token": Object {
                "end": 3,
                "precedingWhitespace": undefined,
                "raw": "foo",
                "start": 0,
                "text": "foo",
                "type": "IDENTIFIER",
              },
              "type": "token",
            },
          ],
          "hasSemicolon": true,
          "type": "statement",
        },
        Object {
          "children": Array [
            Object {
              "token": Object {
                "end": 8,
                "precedingWhitespace": " ",
                "raw": "bar",
                "start": 5,
                "text": "bar",
                "type": "IDENTIFIER",
              },
              "type": "token",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses function call', () => {
    expect(parse('SELECT SQRT(2)')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "nameToken": Object {
                    "end": 11,
                    "precedingWhitespace": " ",
                    "raw": "SQRT",
                    "start": 7,
                    "text": "SQRT",
                    "type": "RESERVED_FUNCTION_NAME",
                  },
                  "parenthesis": Object {
                    "children": Array [
                      Object {
                        "token": Object {
                          "end": 13,
                          "precedingWhitespace": undefined,
                          "raw": "2",
                          "start": 12,
                          "text": "2",
                          "type": "NUMBER",
                        },
                        "type": "token",
                      },
                    ],
                    "closeParen": ")",
                    "openParen": "(",
                    "type": "parenthesis",
                  },
                  "type": "function_call",
                },
              ],
              "nameToken": Object {
                "end": 6,
                "precedingWhitespace": undefined,
                "raw": "SELECT",
                "start": 0,
                "text": "SELECT",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses array subscript', () => {
    expect(parse('SELECT my_array[OFFSET(5)]')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "arrayToken": Object {
                    "end": 15,
                    "precedingWhitespace": " ",
                    "raw": "my_array",
                    "start": 7,
                    "text": "my_array",
                    "type": "IDENTIFIER",
                  },
                  "parenthesis": Object {
                    "children": Array [
                      Object {
                        "nameToken": Object {
                          "end": 22,
                          "precedingWhitespace": undefined,
                          "raw": "OFFSET",
                          "start": 16,
                          "text": "OFFSET",
                          "type": "RESERVED_FUNCTION_NAME",
                        },
                        "parenthesis": Object {
                          "children": Array [
                            Object {
                              "token": Object {
                                "end": 24,
                                "precedingWhitespace": undefined,
                                "raw": "5",
                                "start": 23,
                                "text": "5",
                                "type": "NUMBER",
                              },
                              "type": "token",
                            },
                          ],
                          "closeParen": ")",
                          "openParen": "(",
                          "type": "parenthesis",
                        },
                        "type": "function_call",
                      },
                    ],
                    "closeParen": "]",
                    "openParen": "[",
                    "type": "parenthesis",
                  },
                  "type": "array_subscript",
                },
              ],
              "nameToken": Object {
                "end": 6,
                "precedingWhitespace": undefined,
                "raw": "SELECT",
                "start": 0,
                "text": "SELECT",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses parenthesized expressions', () => {
    expect(parse('SELECT (birth_year - (CURRENT_DATE + 1))')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "token": Object {
                        "end": 18,
                        "precedingWhitespace": undefined,
                        "raw": "birth_year",
                        "start": 8,
                        "text": "birth_year",
                        "type": "IDENTIFIER",
                      },
                      "type": "token",
                    },
                    Object {
                      "token": Object {
                        "end": 20,
                        "precedingWhitespace": " ",
                        "raw": "-",
                        "start": 19,
                        "text": "-",
                        "type": "OPERATOR",
                      },
                      "type": "token",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "token": Object {
                            "end": 34,
                            "precedingWhitespace": undefined,
                            "raw": "CURRENT_DATE",
                            "start": 22,
                            "text": "CURRENT_DATE",
                            "type": "IDENTIFIER",
                          },
                          "type": "token",
                        },
                        Object {
                          "token": Object {
                            "end": 36,
                            "precedingWhitespace": " ",
                            "raw": "+",
                            "start": 35,
                            "text": "+",
                            "type": "OPERATOR",
                          },
                          "type": "token",
                        },
                        Object {
                          "token": Object {
                            "end": 38,
                            "precedingWhitespace": " ",
                            "raw": "1",
                            "start": 37,
                            "text": "1",
                            "type": "NUMBER",
                          },
                          "type": "token",
                        },
                      ],
                      "closeParen": ")",
                      "openParen": "(",
                      "type": "parenthesis",
                    },
                  ],
                  "closeParen": ")",
                  "openParen": "(",
                  "type": "parenthesis",
                },
              ],
              "nameToken": Object {
                "end": 6,
                "precedingWhitespace": undefined,
                "raw": "SELECT",
                "start": 0,
                "text": "SELECT",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses BETWEEN expression', () => {
    expect(parse('WHERE age BETWEEN 10 and 15;')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "token": Object {
                    "end": 9,
                    "precedingWhitespace": " ",
                    "raw": "age",
                    "start": 6,
                    "text": "age",
                    "type": "IDENTIFIER",
                  },
                  "type": "token",
                },
                Object {
                  "andToken": Object {
                    "end": 24,
                    "precedingWhitespace": " ",
                    "raw": "and",
                    "start": 21,
                    "text": "AND",
                    "type": "RESERVED_LOGICAL_OPERATOR",
                  },
                  "betweenToken": Object {
                    "end": 17,
                    "precedingWhitespace": " ",
                    "raw": "BETWEEN",
                    "start": 10,
                    "text": "BETWEEN",
                    "type": "RESERVED_KEYWORD",
                  },
                  "expr1": Object {
                    "end": 20,
                    "precedingWhitespace": " ",
                    "raw": "10",
                    "start": 18,
                    "text": "10",
                    "type": "NUMBER",
                  },
                  "expr2": Object {
                    "end": 27,
                    "precedingWhitespace": " ",
                    "raw": "15",
                    "start": 25,
                    "text": "15",
                    "type": "NUMBER",
                  },
                  "type": "between_predicate",
                },
              ],
              "nameToken": Object {
                "end": 5,
                "precedingWhitespace": undefined,
                "raw": "WHERE",
                "start": 0,
                "text": "WHERE",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": true,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses LIMIT clause', () => {
    expect(parse('LIMIT 10')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "count": Array [
                Object {
                  "token": Object {
                    "end": 8,
                    "precedingWhitespace": " ",
                    "raw": "10",
                    "start": 6,
                    "text": "10",
                    "type": "NUMBER",
                  },
                  "type": "token",
                },
              ],
              "limitToken": Object {
                "end": 5,
                "precedingWhitespace": undefined,
                "raw": "LIMIT",
                "start": 0,
                "text": "LIMIT",
                "type": "RESERVED_COMMAND",
              },
              "type": "limit_clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses LIMIT clause with offset', () => {
    expect(parse('LIMIT 200, 10')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "count": Array [
                Object {
                  "token": Object {
                    "end": 13,
                    "precedingWhitespace": " ",
                    "raw": "10",
                    "start": 11,
                    "text": "10",
                    "type": "NUMBER",
                  },
                  "type": "token",
                },
              ],
              "limitToken": Object {
                "end": 5,
                "precedingWhitespace": undefined,
                "raw": "LIMIT",
                "start": 0,
                "text": "LIMIT",
                "type": "RESERVED_COMMAND",
              },
              "offset": Array [
                Object {
                  "token": Object {
                    "end": 9,
                    "precedingWhitespace": " ",
                    "raw": "200",
                    "start": 6,
                    "text": "200",
                    "type": "NUMBER",
                  },
                  "type": "token",
                },
              ],
              "type": "limit_clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses SELECT *', () => {
    expect(parse('SELECT *')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "all_columns_asterisk",
                },
              ],
              "nameToken": Object {
                "end": 6,
                "precedingWhitespace": undefined,
                "raw": "SELECT",
                "start": 0,
                "text": "SELECT",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });

  it('parses SELECT DISTINCT *', () => {
    expect(parse('SELECT DISTINCT *')).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "all_columns_asterisk",
                },
              ],
              "nameToken": Object {
                "end": 15,
                "precedingWhitespace": undefined,
                "raw": "SELECT DISTINCT",
                "start": 0,
                "text": "SELECT DISTINCT",
                "type": "RESERVED_COMMAND",
              },
              "type": "clause",
            },
          ],
          "hasSemicolon": false,
          "type": "statement",
        },
      ]
    `);
  });
});
