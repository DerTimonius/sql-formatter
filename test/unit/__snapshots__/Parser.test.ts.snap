// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parses BETWEEN expression 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 9,
              "precedingWhitespace": " ",
              "raw": "age",
              "start": 6,
              "text": "age",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
          Object {
            "andToken": Object {
              "end": 24,
              "precedingWhitespace": " ",
              "raw": "AND",
              "start": 21,
              "text": "AND",
              "type": "AND",
            },
            "betweenToken": Object {
              "end": 17,
              "precedingWhitespace": " ",
              "raw": "BETWEEN",
              "start": 10,
              "text": "BETWEEN",
              "type": "BETWEEN",
            },
            "expr1": Array [
              Object {
                "token": Object {
                  "end": 20,
                  "precedingWhitespace": " ",
                  "raw": "18",
                  "start": 18,
                  "text": "18",
                  "type": "NUMBER",
                },
                "type": "token",
              },
            ],
            "expr2": Array [
              Object {
                "token": Object {
                  "end": 27,
                  "precedingWhitespace": " ",
                  "raw": "63",
                  "start": 25,
                  "text": "63",
                  "type": "NUMBER",
                },
                "type": "token",
              },
            ],
            "type": "between_predicate",
          },
        ],
        "nameToken": Object {
          "end": 5,
          "precedingWhitespace": undefined,
          "raw": "WHERE",
          "start": 0,
          "text": "WHERE",
          "type": "RESERVED_COMMAND",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with count 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "token": Object {
              "end": 8,
              "precedingWhitespace": " ",
              "raw": "15",
              "start": 6,
              "text": "15",
              "type": "NUMBER",
            },
            "type": "token",
          },
        ],
        "limitToken": Object {
          "end": 5,
          "precedingWhitespace": undefined,
          "raw": "LIMIT",
          "start": 0,
          "text": "LIMIT",
          "type": "LIMIT",
        },
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with longer expressions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "token": Object {
              "end": 16,
              "precedingWhitespace": " ",
              "raw": "3",
              "start": 15,
              "text": "3",
              "type": "NUMBER",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 18,
              "precedingWhitespace": " ",
              "raw": "*",
              "start": 17,
              "text": "*",
              "type": "ASTERISK",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 20,
              "precedingWhitespace": " ",
              "raw": "2",
              "start": 19,
              "text": "2",
              "type": "NUMBER",
            },
            "type": "token",
          },
        ],
        "limitToken": Object {
          "end": 5,
          "precedingWhitespace": undefined,
          "raw": "LIMIT",
          "start": 0,
          "text": "LIMIT",
          "type": "LIMIT",
        },
        "offset": Array [
          Object {
            "token": Object {
              "end": 8,
              "precedingWhitespace": " ",
              "raw": "50",
              "start": 6,
              "text": "50",
              "type": "NUMBER",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 10,
              "precedingWhitespace": " ",
              "raw": "+",
              "start": 9,
              "text": "+",
              "type": "OPERATOR",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 13,
              "precedingWhitespace": " ",
              "raw": "50",
              "start": 11,
              "text": "50",
              "type": "NUMBER",
            },
            "type": "token",
          },
        ],
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with offset and count 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "token": Object {
              "end": 13,
              "precedingWhitespace": " ",
              "raw": "15",
              "start": 11,
              "text": "15",
              "type": "NUMBER",
            },
            "type": "token",
          },
        ],
        "limitToken": Object {
          "end": 5,
          "precedingWhitespace": undefined,
          "raw": "LIMIT",
          "start": 0,
          "text": "LIMIT",
          "type": "LIMIT",
        },
        "offset": Array [
          Object {
            "token": Object {
              "end": 9,
              "precedingWhitespace": " ",
              "raw": "100",
              "start": 6,
              "text": "100",
              "type": "NUMBER",
            },
            "type": "token",
          },
        ],
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses SELECT * 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "all_columns_asterisk",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses array subscript 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "arrayToken": Object {
              "end": 15,
              "precedingWhitespace": " ",
              "raw": "my_array",
              "start": 7,
              "text": "my_array",
              "type": "ARRAY_IDENTIFIER",
            },
            "parenthesis": Object {
              "children": Array [
                Object {
                  "token": Object {
                    "end": 17,
                    "precedingWhitespace": undefined,
                    "raw": "5",
                    "start": 16,
                    "text": "5",
                    "type": "NUMBER",
                  },
                  "type": "token",
                },
              ],
              "closeParen": "]",
              "openParen": "[",
              "type": "parenthesis",
            },
            "type": "array_subscript",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses curly braces 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "token": Object {
                  "end": 11,
                  "precedingWhitespace": undefined,
                  "raw": "foo",
                  "start": 8,
                  "text": "foo",
                  "type": "IDENTIFIER",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 12,
                  "precedingWhitespace": undefined,
                  "raw": ":",
                  "start": 11,
                  "text": ":",
                  "type": "OPERATOR",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 16,
                  "precedingWhitespace": " ",
                  "raw": "bar",
                  "start": 13,
                  "text": "bar",
                  "type": "IDENTIFIER",
                },
                "type": "token",
              },
            ],
            "closeParen": "}",
            "openParen": "{",
            "type": "parenthesis",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses function call 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "nameToken": Object {
              "end": 11,
              "precedingWhitespace": " ",
              "raw": "sqrt",
              "start": 7,
              "text": "SQRT",
              "type": "RESERVED_FUNCTION_NAME",
            },
            "parenthesis": Object {
              "children": Array [
                Object {
                  "token": Object {
                    "end": 13,
                    "precedingWhitespace": undefined,
                    "raw": "2",
                    "start": 12,
                    "text": "2",
                    "type": "NUMBER",
                  },
                  "type": "token",
                },
              ],
              "closeParen": ")",
              "openParen": "(",
              "type": "parenthesis",
            },
            "type": "function_call",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses function name with and without parameters 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 19,
              "precedingWhitespace": " ",
              "raw": "CURRENT_TIME",
              "start": 7,
              "text": "CURRENT_TIME",
              "type": "RESERVED_KEYWORD",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 21,
              "precedingWhitespace": " ",
              "raw": "a",
              "start": 20,
              "text": "a",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
          Object {
            "token": Object {
              "end": 22,
              "precedingWhitespace": undefined,
              "raw": ",",
              "start": 21,
              "text": ",",
              "type": "COMMA",
            },
            "type": "token",
          },
          Object {
            "nameToken": Object {
              "end": 35,
              "precedingWhitespace": " ",
              "raw": "CURRENT_TIME",
              "start": 23,
              "text": "CURRENT_TIME",
              "type": "RESERVED_FUNCTION_NAME",
            },
            "parenthesis": Object {
              "children": Array [],
              "closeParen": ")",
              "openParen": "(",
              "type": "parenthesis",
            },
            "type": "function_call",
          },
          Object {
            "token": Object {
              "end": 39,
              "precedingWhitespace": " ",
              "raw": "b",
              "start": 38,
              "text": "b",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses list of statements 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "token": Object {
          "end": 3,
          "precedingWhitespace": undefined,
          "raw": "foo",
          "start": 0,
          "text": "foo",
          "type": "IDENTIFIER",
        },
        "type": "token",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
  Object {
    "children": Array [
      Object {
        "token": Object {
          "end": 8,
          "precedingWhitespace": " ",
          "raw": "bar",
          "start": 5,
          "text": "bar",
          "type": "IDENTIFIER",
        },
        "type": "token",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses parenthesized expressions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "token": Object {
                  "end": 18,
                  "precedingWhitespace": undefined,
                  "raw": "birth_year",
                  "start": 8,
                  "text": "birth_year",
                  "type": "IDENTIFIER",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 20,
                  "precedingWhitespace": " ",
                  "raw": "-",
                  "start": 19,
                  "text": "-",
                  "type": "OPERATOR",
                },
                "type": "token",
              },
              Object {
                "children": Array [
                  Object {
                    "token": Object {
                      "end": 34,
                      "precedingWhitespace": undefined,
                      "raw": "CURRENT_DATE",
                      "start": 22,
                      "text": "CURRENT_DATE",
                      "type": "IDENTIFIER",
                    },
                    "type": "token",
                  },
                  Object {
                    "token": Object {
                      "end": 36,
                      "precedingWhitespace": " ",
                      "raw": "+",
                      "start": 35,
                      "text": "+",
                      "type": "OPERATOR",
                    },
                    "type": "token",
                  },
                  Object {
                    "token": Object {
                      "end": 38,
                      "precedingWhitespace": " ",
                      "raw": "1",
                      "start": 37,
                      "text": "1",
                      "type": "NUMBER",
                    },
                    "type": "token",
                  },
                ],
                "closeParen": ")",
                "openParen": "(",
                "type": "parenthesis",
              },
            ],
            "closeParen": ")",
            "openParen": "(",
            "type": "parenthesis",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses set operations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 10,
              "precedingWhitespace": " ",
              "raw": "foo",
              "start": 7,
              "text": "foo",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 19,
              "precedingWhitespace": " ",
              "raw": "bar",
              "start": 16,
              "text": "bar",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
        ],
        "nameToken": Object {
          "end": 15,
          "precedingWhitespace": " ",
          "raw": "FROM",
          "start": 11,
          "text": "FROM",
          "type": "RESERVED_COMMAND",
        },
        "type": "clause",
      },
      Object {
        "children": Array [],
        "nameToken": Object {
          "end": 29,
          "precedingWhitespace": " ",
          "raw": "UNION ALL",
          "start": 20,
          "text": "UNION ALL",
          "type": "RESERVED_SET_OPERATION",
        },
        "type": "set_operation",
      },
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 40,
              "precedingWhitespace": " ",
              "raw": "foo",
              "start": 37,
              "text": "foo",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
        ],
        "nameToken": Object {
          "end": 36,
          "precedingWhitespace": " ",
          "raw": "SELECT",
          "start": 30,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
      Object {
        "children": Array [
          Object {
            "token": Object {
              "end": 49,
              "precedingWhitespace": " ",
              "raw": "baz",
              "start": 46,
              "text": "baz",
              "type": "IDENTIFIER",
            },
            "type": "token",
          },
        ],
        "nameToken": Object {
          "end": 45,
          "precedingWhitespace": " ",
          "raw": "FROM",
          "start": 41,
          "text": "FROM",
          "type": "RESERVED_COMMAND",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses square brackets 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "token": Object {
                  "end": 9,
                  "precedingWhitespace": undefined,
                  "raw": "1",
                  "start": 8,
                  "text": "1",
                  "type": "NUMBER",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 10,
                  "precedingWhitespace": undefined,
                  "raw": ",",
                  "start": 9,
                  "text": ",",
                  "type": "COMMA",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 12,
                  "precedingWhitespace": " ",
                  "raw": "2",
                  "start": 11,
                  "text": "2",
                  "type": "NUMBER",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 13,
                  "precedingWhitespace": undefined,
                  "raw": ",",
                  "start": 12,
                  "text": ",",
                  "type": "COMMA",
                },
                "type": "token",
              },
              Object {
                "token": Object {
                  "end": 15,
                  "precedingWhitespace": " ",
                  "raw": "3",
                  "start": 14,
                  "text": "3",
                  "type": "NUMBER",
                },
                "type": "token",
              },
            ],
            "closeParen": "]",
            "openParen": "[",
            "type": "parenthesis",
          },
        ],
        "nameToken": Object {
          "end": 6,
          "precedingWhitespace": undefined,
          "raw": "SELECT",
          "start": 0,
          "text": "SELECT",
          "type": "RESERVED_SELECT",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;
